pkg/server/server.go

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"syscall"
+
+	"github.com/go-logr/logr"
+	"github.com/oklog/run"
+
+	"github.com/NVIDIA/knavigator/pkg/config"
+	"github.com/NVIDIA/knavigator/pkg/engine"
+)
+
+type Server struct {
+	s   *http.Server
+	log *logr.Logger
+}
+
+type TaskHandler struct {
+	eng *engine.Eng
+}
+
+type ResetHandler struct {
+	eng *engine.Eng
+}
+
+func New(log *logr.Logger, eng *engine.Eng, port int) *Server {
+	mux := http.NewServeMux()
+	mux.Handle("/task", &TaskHandler{eng: eng})
+	mux.Handle("/reset", &ResetHandler{eng: eng})
+	return &Server{
+		log: log,
+		s: &http.Server{ // #nosec G112
+			Addr:    fmt.Sprintf(":%d", port),
+			Handler: mux,
+		},
+	}
+}
+
+func (srv *Server) Run() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var g run.Group
+	// Signal handler
+	g.Add(run.SignalHandler(ctx, os.Interrupt, syscall.SIGTERM))
+	// Server
+	g.Add(
+		func() error {
+			srv.log.Info("Starting server", "address", srv.s.Addr)
+			return srv.s.ListenAndServe()
+		},
+		func(err error) {
+			srv.log.Error(err, "Stopping server")
+			if err := srv.s.Shutdown(ctx); err != nil {
+				srv.log.Error(err, "Error during server shutdown")
+			}
+			srv.log.Info("Server stopped")
+		})
+
+	return g.Run()
+}
+
+func (h *TaskHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```

Comment:
I think the server should accept the complete test YAML (https://github.com/NVIDIA/knavigator/blob/main/resources/tests/k8s/test-failed-job.yml) so that the user can send the file via curl to the server. It will make it easy to use. What do you think?

Reply:
agreed, done